### PR TITLE
Form CUDNN graphs recursively and propagrate transposes and layouts

### PR DESCRIPTION
**Pull request by frgossen**

### CODE REVIEW

**Review by ezhulenev — APPROVED** [review]

File: compiler/src/openxla/compiler/nvgpu/Transforms/ConvertHLOToCUDNN.cpp [hunk 1]
```diff
@@ -104,11 +65,112 @@ class CudnnHandleCache {
...
+// Helper functions for permutations.
+
+static SmallVector<int64_t> GetTensorLayoutPermutation(
+    ArrayRef<int64_t> srcSpatialDims, int64_t srcBatchDim,
+    int64_t srcFeatureDim, Layout dst) {
+  assert(srcSpatialDims.size() == 2 && "expect 2 spatial dims");
+  switch (dst) {
+    case Layout::NHWC:
+      return {
+          {srcBatchDim, srcSpatialDims[0], srcSpatialDims[1], srcFeatureDim}};
```
[the remark this answers is not in the document]
Comment (frgossen) [reply compiler/src/openxla/compiler/nvgpu/Transforms/ConvertHLOToCUDNN.cpp:77]: Wasn't on purpose and somehow worked. Fixed.

File: compiler/src/openxla/compiler/nvgpu/Transforms/ConvertHLOToCUDNN.cpp [hunk 2]
```diff
@@ -104,11 +65,112 @@ class CudnnHandleCache {
...
+  }
+}
+
+static SmallVector<int64_t> GetKernelLayoutPermutation(
+    ArrayRef<int64_t> srcSpatialDims, int64_t srcOutputFeatureDim,
+    int64_t srcInputFeatureDim, Layout dst) {
+  assert(srcSpatialDims.size() == 2 && "expect 2 spatial dims");
+  switch (dst) {
+    case Layout::KHWC:
+      return {{srcOutputFeatureDim, srcSpatialDims[0], srcSpatialDims[1],
```
[the remark this answers is not in the document]
Comment (frgossen) [reply compiler/src/openxla/compiler/nvgpu/Transforms/ConvertHLOToCUDNN.cpp:92]: Done.

File: compiler/src/openxla/compiler/nvgpu/Transforms/ConvertHLOToCUDNN.cpp [hunk 3]
```diff
@@ -104,11 +65,112 @@ class CudnnHandleCache {
 };
 }  // namespace
 
-/// Outline HLO ops into CUDNN graphs.
+// Helper functions for permutations.
+
+static SmallVector<int64_t> GetTensorLayoutPermutation(
```
[the remark this answers is not in the document]
Comment (frgossen) [reply compiler/src/openxla/compiler/nvgpu/Transforms/ConvertHLOToCUDNN.cpp:70]: Done,

File: compiler/src/openxla/compiler/nvgpu/Transforms/ConvertHLOToCUDNN.cpp [hunk 4]
```diff
@@ -104,11 +65,112 @@ class CudnnHandleCache {
...
+    case Layout::KCHW:
+      return {{srcOutputFeatureDim, srcInputFeatureDim, srcSpatialDims[0],
+               srcSpatialDims[1]}};
+    default:
+      llvm_unreachable("not a kernel layout");
+  }
+}
+
+static SmallVector<int64_t> GetReversePermutation(
+    const SmallVector<int64_t> &permutation) {
```
[the remark this answers is not in the document]
Comment (frgossen) [reply compiler/src/openxla/compiler/nvgpu/Transforms/ConvertHLOToCUDNN.cpp:103]: Done.

**Pull request merged**